spec/lib/search_logic_spec.rb

```diff
@@ -0,0 +1,79 @@
+RSpec.describe SearchLogic do
+    it "saves new attractions" do
+        geocode_venice
+        
+        places_stub = stub_request(:get, venice_request_url("tourism.attraction")).to_return(status: 200, body: File.read("spec/responses/attractions_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "attractions")
+            expect(result).to all(be_a(Attraction))
+        }.to change(City, :count).by(1).and change(Attraction, :count).by(20)
+
+        expect(geocode_venice).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+
+    it "saves new hotels" do
+        geocode_venice
+
+        places_stub = stub_request(:get, venice_request_url("accommodation.hotel")).to_return(status: 200, body: File.read("spec/responses/hotels_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "hotels")
+            expect(result).to all(be_a(Hotel))
+        }.to change(City, :count).by(1).and change(Hotel, :count).by(20)
+
+        expect(geocode_venice).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+    
+    it "saves new restaurants" do
+        geocode_venice
+
+        places_stub = stub_request(:get, venice_request_url("catering.restaurant")).to_return(status: 200, body: File.read("spec/responses/restaurants_venice.json"))
+
+        expect {
+            result = SearchLogic.get_places_from_city_string("Venice, Italy", "restaurants")
+            expect(result).to all(be_a(Restaurant))
+        }.to change(City, :count).by(1).and change(Restaurant, :count).by(20)
+
+        expect(geocode_venice).to have_been_requested.times(1)
+        expect(places_stub).to have_been_requested.times(1)
+    end
+
+    it "does not create a city if it is already in the database" do
+        places_stub = stub_request(:get, venice_request_url("tourism.attraction")).to_return(status: 200, body: File.read("spec/responses/attractions_venice.json"))
+
+        city = create_venice
+
+        expect {
+            SearchLogic.get_places_from_city_string(city.name, "attractions")
+        }.to change(City, :count).by(0)
+
+        expect(geocode_venice).to have_been_requested.times(0)
```

Comment:
I still need the stub because I expect the test to still call the Places API to create the places

Reply:
Doesn't the `.times(0)` mean it shouldn't be called?

Reply:
Hm no...Geocode is for the city. Places API is a different one. So here I guess I should make clear in the test that Places API should still be called. Sorry about that

Reply:
It's already clear that this scenario requires the places API stub because of the line below. The `geocode_venice` line here, however, is redundant when you have WebMock installed.